Fix comment style when placed after if statement test

jackyho112 opened the pull request:
This is an attempt to fix https://github.com/prettier/prettier/issues/2650

so that comments placed before a test in an if statement with no bracket on the consequent such as 

```
if (typeof err === "string") // comment
    callback(err);
```

will be formatted to 

```
if (typeof err === "string")
    // comment
    callback(err);
```

Let me know if there is any improvement you have in mind. :smile:

vjeux commented on src/comments.js:
```diff
@@ -480,6 +480,13 @@ function handleIfStatementComments(
 
+  // For `if (a) /* comment */ {}`,
+  // we look at the previous character to see if it is a ).
+  if (getPreviousNonSpaceCharacter(text, comment) === ")") {
```
Why not look at the next character and see if it's a `{` ?

jackyho112 replied:
If I check `{`, it would still miss cases such as 

```
if (typeof err === "string") // comment
    callback(err);
```

though.

vjeux commented on src/util.js:
```diff
@@ -198,6 +198,18 @@ function getNextNonSpaceNonCommentCharacter(text, node) {
+    oldIdx = idx;
+    idx = skipSpaces(text, idx, traversalOptions);
+    idx = skipNewline(text, idx, traversalOptions);
+  }
```
You don't handle comments. This is tricky to get right, so let's see if we can make it work with getNextNonSpaceCharacter instead

jackyho112 commented:
Hey @vjeux,

thanks for the suggestions. Based on them, I changed `getPreviousNonSpaceNonCommentCharacter` and added tests that have inline comments before a trailing comment.

`getPreviousNonSpaceNonCommentCharacter` now checks for inline comments. I don't think it's necessary for it to check trailing comments because it would turn the whole thing into one trailing comment block. 

Let me know if there is any other improvement you have in mind.

jackyho112 commented:
@vjeux 

Also, let me know my approach is ok. If not and I were to use `getNextNonSpaceCharacter`, I believe I could check for either a new line or `{`(there could other edge cases but I can try my best to figure it out).

jackyho112 commented on src/util.js:
```diff
@@ -216,6 +219,8 @@ function getNextNonSpaceNonCommentCharacter(text, node) {
 
+// This function doesn't ignore trailing comments unlike
+// getNextNonSpaceNonCommentCharacter.
 function getPreviousNonSpaceNonCommentCharacter(text, node) {
```
I tried to use `getNextNonSpaceNonCommentCharacter` but couldn't find a way to make it work perfectly to solve this issue.

Kind like `isPreviousLineEmpty`, this doesn't skip trailing comments but should cover the use cases for this issue because any previous trailing comment will just turn the whole node on that line to be a trailing comment. Please let me know if that is ok.

vjeux commented:
This solution looks a lot cleaner, thanks for working on it :)
